src/main/scala/mesosphere/mesos/TaskGroupBuilder.scala

```diff
@@ -0,0 +1,394 @@
+    builder.addResources(scalarResource("mem", container.resources.mem))
+    builder.addResources(scalarResource("disk", container.resources.disk))
+    builder.addResources(scalarResource("gpus", container.resources.gpus.toDouble))
+
+    val labels = podDefinition.labels ++ container.labels.map(_.values).getOrElse(Map.empty[String, String])
+
+    if (labels.nonEmpty)
+      builder.setLabels(mesos.Labels.newBuilder.addAllLabels(labels.map {
+        case (key, value) =>
+          mesos.Label.newBuilder.setKey(key).setValue(value).build
+      }.asJava))
+
+    val commandInfo = computeCommandInfo(
+      podDefinition,
+      instanceId,
+      container,
+      offer.getHostname,
+      portsEnvVars)
+
+    builder.setCommand(commandInfo)
+
+    val containerInfo = computeContainerInfo(podDefinition.podVolumes, container)
+
+    builder.setContainer(containerInfo)
+
+    container.healthCheck.foreach { healthCheck =>
+      builder.setHealthCheck(computeHealthCheck(healthCheck, container.endpoints))
+    }
+
+    builder
+  }
+
+  private[this] def computeExecutorInfo(
+    podDefinition: PodDefinition,
+    portsMatch: PortsMatch,
+    instanceId: Instance.Id): mesos.ExecutorInfo.Builder = {
+    val executorID = mesos.ExecutorID.newBuilder.setValue(f"marathon-${instanceId.idString}")
+
+    val executorInfo = mesos.ExecutorInfo.newBuilder
+      .setType(mesos.ExecutorInfo.Type.DEFAULT)
+      .setExecutorId(executorID)
+
+    executorInfo.addResources(scalarResource("cpus", PodDefinition.DefaultExecutorCpus))
+    executorInfo.addResources(scalarResource("mem", PodDefinition.DefaultExecutorMem))
+    executorInfo.addAllResources(portsMatch.resources.asJava)
+
+    if (podDefinition.networks.nonEmpty) {
+      val containerInfo = mesos.ContainerInfo.newBuilder
+        .setType(mesos.ContainerInfo.Type.MESOS)
+
+      // TODO: Does 'DiscoveryInfo' need to be set?
+
+      podDefinition.networks
+        .filter(network => network.name.isDefined && network.mode == raml.NetworkMode.Container)
+        .foreach{ containerNetwork =>
+          val networkInfo = mesos.NetworkInfo.newBuilder
+            .setName(containerNetwork.name.get)
+
+          containerNetwork.labels.foreach { labels =>
+            networkInfo.setLabels(
+              mesos.Labels.newBuilder.addAllLabels(labels.values.map {
+                case (key, value) =>
+                  mesos.Label.newBuilder.setKey(key).setValue(value).build
+              }.asJava))
+          }
+
+          // TODO: add port mappings here
+
+          containerInfo.addNetworkInfos(networkInfo)
+        }
+
+      executorInfo.setContainer(containerInfo)
+    }
+
+    if (podDefinition.labels.nonEmpty) {
+      val labels = mesos.Labels.newBuilder
+        .addAllLabels(podDefinition.labels.map {
+          case (key, value) =>
+            mesos.Label.newBuilder.setKey(key).setValue(value).build
+        }.asJava)
+
+      executorInfo.setLabels(labels)
+    }
```

Comment:
Think I even found something nicer. Factored out a function that does `Map[String, String] => mesos.Labels`.